internal/download/get.go

```diff
@@ -31,81 +34,115 @@ func init() {
 
-func httpGetWithProgress(url string, progress *progress.Progress) ([]byte, *failures.Failure) {
-	logging.Debug("Retrieving url: %s", url)
-	client := retryhttp.NewClient(0 /* 0 = no timeout */, 5)
-	resp, err := client.Get(url)
-	if err != nil {
-		code := -1
-		if resp != nil {
-			code = resp.StatusCode
-		}
-		return nil, failures.FailNetwork.Wrap(err, locale.Tl("err_network_get", "Status code: {{.V0}}", strconv.Itoa(code)))
-	}
-	defer resp.Body.Close()
+	s3m := parseS3URL(url)
 
-	if resp.StatusCode != 200 {
-		return nil, failures.FailNetwork.New("err_invalid_status_code", strconv.Itoa(resp.StatusCode))
+	// Prepare AWS session
+	sess, err := session.NewSession(&aws.Config{
+		Region:      aws.String(s3m.Region),
+		Credentials: credentials.AnonymousCredentials,
+	})
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not create aws session")
 	}
 
-	var total int
-	length := resp.Header.Get("Content-Length")
-	if length == "" {
-		total = 1
+	// Grab file size
+	var length int64 = 0
+	res, err := http.Get(url.String())
+	if err != nil {
+		logging.Debug("Could not grab url: %v", err)
+	} else if res.StatusCode != http.StatusOK {
+		logging.Debug("Could not grab url due to statuscode: %d", res.StatusCode)
 	} else {
-		total, err = strconv.Atoi(length)
+		lengthInt, err := strconv.Atoi(res.Header.Get("Content-Length"))
 		if err != nil {
-			logging.Debug("Content-length: %v", length)
-			return nil, failures.FailInput.Wrap(err)
+			logging.Debug("Could not grab content-length: %v", err)
+		} else {
+			length = int64(lengthInt)
 		}
 	}
 
-	bar := progress.AddByteProgressBar(int64(total))
-
-	src := resp.Body
-	var dst bytes.Buffer
+	// Close early cause we're just looking at the header
+	// Yes normally you'd use a HEAD for this, but S3 presigned URLs don't support HEAD requests
+	res.Body.Close()
 
-	src = bar.ProxyReader(resp.Body)
+	// Record progress
+	bar := progress.AddByteProgressBar(length)
+	cb := func(length int) {
+		if !bar.Completed() {
+			// Failsafe, so we don't get blocked by a progressbar
+			bar.IncrBy(length)
+		}
+	}
 
-	_, err = io.Copy(&dst, src)
+	// Prepare result recorder
+	b := []byte{}
+	w := NewWriteAtBuffer(b, cb)
+
+	dl := s3manager.NewDownloader(sess)
+	dl.RequestOptions = append(dl.RequestOptions, func(r *request.Request) {
+		r.Handlers.Build.PushBack(func(r *request.Request) {
+			// Work around AWS rewriting our query in the wrong order, causing signing to fail
+			r.HTTPRequest.URL.RawQuery = url.RawQuery
+		})
+	})
+	_, err = dl.Download(w,
+		&s3.GetObjectInput{
+			Bucket: aws.String(s3m.Bucket),
+			Key:    aws.String(s3m.Key),
+		})
 	if err != nil {
-		return nil, failures.FailInput.Wrap(err)
+		return nil, locale.WrapError(err, "err_dl_s3", "Downloading failed due to underlying S3 error: {{.V0}}.", err.Error())
 	}
 
-	if !bar.Completed() {
-		// Failsafe, so we don't get blocked by a progressbar
-		bar.IncrBy(total)
-	}
+	bar.Abort(true) // ensure we don't get stuck on an incomplete bar
```

Comment:
note: This will interrupt rather than update to 100%.

Reply:
In theory it should "interrupt" when it's at 100% anyway. This is just a failsafe. Without this if for some reason it doesn't reach 100% due to some unforeseen bug it will just hang and the process won't exit.